### 2.1 Import Data and Required Packages
####  Importing Pandas, Numpy, Matplotlib, Seaborn and Warings Library.

In [5]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings('ignore')

#### Import the CSV Data as Pandas DataFrame

In [6]:
df = pd.read_csv('data/raw_data.csv')

#### Show Top 5 Records

In [4]:
df.head()

id  gender      customer_type  age   type_of_travel     Class  \
0   70172    Male     Loyal Customer   13  Personal Travel  Eco Plus   
1    5047    Male  disloyal Customer   25  Business travel  Business   
2  110028  Female     Loyal Customer   26  Business travel  Business   
3   24026  Female     Loyal Customer   25  Business travel  Business   
4  119299    Male     Loyal Customer   61  Business travel  Business   

   flight_distance  inflight_wifi_service  departure_arrival_time_convenient  \
0              460                      3                                  4   
1              235                      3                                  2   
2             1142                      2                                  2   
3              562                      2                                  5   
4              214                      3                                  3   

   ease_of_online_booking  ...  inflight_entertainment  on_board_service  \
0                       3  ...                       5                 4   
1                       3  ...                       1                 1   
2                       2  ...                       5                 4   
3                       5  ...                       2                 2   
4                       3  ...                       3                 3   

   leg_room_service  baggage_handling  checkin_service  inflight_service  \
0                 3                 4                4                 5   
1                 5                 3                1                 4   
2                 3                 4                4                 4   
3                 5                 3                1                 4   
4                 4                 4                3                 3   

   cleanliness  departure_delay_in_minutes  arrival_delay_in_minutes  \
0            5                          25                      18.0   
1            1                           1                       6.0   
2            5                           0                       0.0   
3            2                          11                       9.0   
4            3                           0                       0.0   

              satisfaction  
0  neutral or dissatisfied  
1  neutral or dissatisfied  
2                satisfied  
3  neutral or dissatisfied  
4                satisfied  

[5 rows x 24 columns]

#### Shape of the dataset

In [6]:
df.shape

(103904, 24)

### 2.2 Dataset information

- id : unique for each passenger
- gender : sex of passenger  -> (Male/female)
- customer_type : loyality of passenger -> (Loyal Disloyal)
- age : Age of the passenger
- type_of_travel : (Personal/Business)
- Class : (Eco, Eco Plus, Business)
- flight_distance
- inflight_wifi_service : rating from 0 to 5
- departure_arrival_time_convenient : rating from 0 to 5
- ease_of_online_booking : rating from 0 to 5
- gate_location : rating from 0 to 5
- food_and_drink : rating from 0 to 5
- online_boarding : rating from 0 to 5
- seat_comfort : rating from 0 to 5
- inflight_entertainment : rating from 0 to 5
- on_board_service : rating from 0 to 5
- leg_room_service : rating from 0 to 5
- baggage_handling : rating from 0 to 5
- checkin_service : rating from 0 to 5
- inflight_service : rating from 0 to 5
- cleanliness : rating from 0 to 5
- departure_delay_in_minutes
- arrival_delay_in_minutes
- satisfaction : satisfied or neutral/dissatisfied

### 3. Data Checks to perform

- Check Missing values
- Check Duplicates
- Check data type
- Check the number of unique values of each column
- Check statistics of data set
- Check various categories present in the different categorical column

In [13]:
df.isna().sum()

id                                     0
gender                                 0
customer_type                          0
age                                    0
type_of_travel                         0
Class                                  0
flight_distance                        0
inflight_wifi_service                  0
departure_arrival_time_convenient      0
ease_of_online_booking                 0
gate_location                          0
food_and_drink                         0
online_boarding                        0
seat_comfort                           0
inflight_entertainment                 0
on_board_service                       0
leg_room_service                       0
baggage_handling                       0
checkin_service                        0
inflight_service                       0
cleanliness                            0
departure_delay_in_minutes             0
arrival_delay_in_minutes             310
satisfaction                           0
dtype: int64

#### There are 310 missing values in 'arrival_delay_in_minutes' column of the data set

### 3.2 Check Duplicates

In [14]:
df.duplicated().sum()

0

#### There are no duplicates  values in the data set

### 3.3 Check data types

In [15]:
# Check Null and Dtypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103904 entries, 0 to 103903
Data columns (total 24 columns):
 #   Column                             Non-Null Count   Dtype  
---  ------                             --------------   -----  
 0   id                                 103904 non-null  int64  
 1   gender                             103904 non-null  object 
 2   customer_type                      103904 non-null  object 
 3   age                                103904 non-null  int64  
 4   type_of_travel                     103904 non-null  object 
 5   Class                              103904 non-null  object 
 6   flight_distance                    103904 non-null  int64  
 7   inflight_wifi_service              103904 non-null  int64  
 8   departure_arrival_time_convenient  103904 non-null  int64  
 9   ease_of_online_booking             103904 non-null  int64  
 10  gate_location                      103904 non-null  int64  
 11  food_and_drink                     1039

### 3.4 Checking the number of unique values of each column

In [16]:
df.nunique()

id                                   103904
gender                                    2
customer_type                             2
age                                      75
type_of_travel                            2
Class                                     3
flight_distance                        3802
inflight_wifi_service                     6
departure_arrival_time_convenient         6
ease_of_online_booking                    6
gate_location                             6
food_and_drink                            6
online_boarding                           6
seat_comfort                              6
inflight_entertainment                    6
on_board_service                          6
leg_room_service                          6
baggage_handling                          5
checkin_service                           6
inflight_service                          6
cleanliness                               6
departure_delay_in_minutes              446
arrival_delay_in_minutes        

### 3.5 Check statistics of data set

In [17]:
df.describe()

id            age  flight_distance  inflight_wifi_service  \
count  103904.000000  103904.000000    103904.000000          103904.000000   
mean    64924.210502      39.379706      1189.448375               2.729683   
std     37463.812252      15.114964       997.147281               1.327829   
min         1.000000       7.000000        31.000000               0.000000   
25%     32533.750000      27.000000       414.000000               2.000000   
50%     64856.500000      40.000000       843.000000               3.000000   
75%     97368.250000      51.000000      1743.000000               4.000000   
max    129880.000000      85.000000      4983.000000               5.000000   

       departure_arrival_time_convenient  ease_of_online_booking  \
count                      103904.000000           103904.000000   
mean                            3.060296                2.756901   
std                             1.525075                1.398929   
min                             0.000000                0.000000   
25%                             2.000000                2.000000   
50%                             3.000000                3.000000   
75%                             4.000000                4.000000   
max                             5.000000                5.000000   

       gate_location  food_and_drink  online_boarding   seat_comfort  \
count  103904.000000   103904.000000    103904.000000  103904.000000   
mean        2.976883        3.202129         3.250375       3.439396   
std         1.277621        1.329533         1.349509       1.319088   
min         0.000000        0.000000         0.000000       0.000000   
25%         2.000000        2.000000         2.000000       2.000000   
50%         3.000000        3.000000         3.000000       4.000000   
75%         4.000000        4.000000         4.000000       5.000000   
max         5.000000        5.000000         5.000000       5.000000   

       inflight_entertainment  on_board_service  leg_room_service  \
count           103904.000000     103904.000000     103904.000000   
mean                 3.358158          3.382363          3.351055   
std                  1.332991          1.288354          1.315605   
min                  0.000000          0.000000          0.000000   
25%                  2.000000          2.000000          2.000000   
50%                  4.000000          4.000000          4.000000   
75%                  4.000000          4.000000          4.000000   
max                  5.000000          5.000000          5.000000   

       baggage_handling  checkin_service  inflight_service    cleanliness  \
count     103904.000000    103904.000000     103904.000000  103904.000000   
mean           3.631833         3.304290          3.640428       3.286351   
std            1.180903         1.265396          1.175663       1.312273   
min            1.000000         0.000000          0.000000       0.000000   
25%            3.000000         3.000000          3.000000       2.000000   
50%            4.000000         3.000000          4.000000       3.000000   
75%            5.000000         4.000000          5.000000       4.000000   
max            5.000000         5.000000          5.000000       5.000000   

       departure_delay_in_minutes  arrival_delay_in_minutes  
count               103904.000000             103594.000000  
mean                    14.815618                 15.178678  
std                     38.230901                 38.698682  
min                      0.000000                  0.000000  
25%                      0.000000                  0.000000  
50%                      0.000000                  0.000000  
75%                     12.000000                 13.000000  
max                   1592.000000               1584.000000

#### Insight
- Average age of the passenger is 39.
- Min age of the Passenger is 7.
- Max age of the Passenger is 85.
- Services with Highest Mean : inflight_service (3.640428)
- Services with lowest Mean : inflight_wifi_service (2.729683)

### 3.7 Exploring Data

In [18]:
for column in df.columns:
    unique_values = df[column].unique()
    print(f"Unique values for column '{column}':")
    print(unique_values)
    print()

Unique values for column 'id':
[ 70172   5047 110028 ...  68825  54173  62567]

Unique values for column 'gender':
['Male' 'Female']

Unique values for column 'customer_type':
['Loyal Customer' 'disloyal Customer']

Unique values for column 'age':
[13 25 26 61 47 52 41 20 24 12 53 33 45 38  9 17 43 58 23 57 49 36 22 31
 15 35 67 37 40 34 39 50 29 54 21 28 27 69 60 48 59 46 30 66 64 44 51 32
 19 42 16 11 62  8 56 68 55 18 65 72 70 63 10  7 14 80 74 71 85 73 76 77
 75 79 78]

Unique values for column 'type_of_travel':
['Personal Travel' 'Business travel']

Unique values for column 'Class':
['Eco Plus' 'Business' 'Eco']

Unique values for column 'flight_distance':
[ 460  235 1142 ...  974 1479  400]

Unique values for column 'inflight_wifi_service':
[3 2 4 1 5 0]

Unique values for column 'departure_arrival_time_convenient':
[4 2 5 3 1 0]

Unique values for column 'ease_of_online_booking':
[3 2 5 4 1 0]

Unique values for column 'gate_location':
[1 3 2 5 4 0]

Unique values for column 'fo

In [20]:
# define numerical & categorical columns
numeric_features = [feature for feature in df.columns if df[feature].dtype != 'O']
categorical_features = [feature for feature in df.columns if df[feature].dtype == 'O']

# print columns
print('We have {} numerical features : {}'.format(len(numeric_features), numeric_features))
print('\nWe have {} categorical features : {}'.format(len(categorical_features), categorical_features))

We have 19 numerical features : ['id', 'age', 'flight_distance', 'inflight_wifi_service', 'departure_arrival_time_convenient', 'ease_of_online_booking', 'gate_location', 'food_and_drink', 'online_boarding', 'seat_comfort', 'inflight_entertainment', 'on_board_service', 'leg_room_service', 'baggage_handling', 'checkin_service', 'inflight_service', 'cleanliness', 'departure_delay_in_minutes', 'arrival_delay_in_minutes']

We have 5 categorical features : ['gender', 'customer_type', 'type_of_travel', 'Class', 'satisfaction']


In [23]:
rating_columns=['inflight_wifi_service','departure_arrival_time_convenient','ease_of_online_booking','gate_location','food_and_drink','online_boarding','seat_comfort','inflight_entertainment','on_board_service','leg_room_service','baggage_handling','checkin_service','inflight_service','cleanliness']
for column in rating_columns:
    max_rating = df[df[column] == 5]['id'].count()
    print(f'Number of Passengers rated 5 for {column} : {max_rating}')
print()
for column in rating_columns:
    min_rating = df[df[column] == 0]['id'].count()
    print(f'Number of Passengers rated 0 for {column} : {min_rating}')

Number of Passengers rated 5 for inflight_wifi_service : 11469
Number of Passengers rated 5 for departure_arrival_time_convenient : 22403
Number of Passengers rated 5 for ease_of_online_booking : 13851
Number of Passengers rated 5 for gate_location : 13879
Number of Passengers rated 5 for food_and_drink : 22313
Number of Passengers rated 5 for online_boarding : 20713
Number of Passengers rated 5 for seat_comfort : 26470
Number of Passengers rated 5 for inflight_entertainment : 25213
Number of Passengers rated 5 for on_board_service : 23648
Number of Passengers rated 5 for leg_room_service : 24667
Number of Passengers rated 5 for baggage_handling : 27131
Number of Passengers rated 5 for checkin_service : 20619
Number of Passengers rated 5 for inflight_service : 27116
Number of Passengers rated 5 for cleanliness : 22689

Number of Passengers rated 0 for inflight_wifi_service : 3103
Number of Passengers rated 0 for departure_arrival_time_convenient : 5300
Number of Passengers rated 0 for 

#### Insight
- baggage_handling service has maximun 5 star rating from passengers
- inflight_wifi_service service has minimum 5 star rating from passengers
- departure_arrival_time_convenient service has maximun 0 star rating from passengers
- baggage_handling has No 0 star rating from passengers

### 4. Exploring Data ( Visualization )
#### 4.1 Visualize average score distribution to make some conclusion. 
- Histogram
- Kernel Distribution Function (KDE)

#### 4.1.1 Histogram & KD
